Applications/Spire/Source/Ui/CalendarDatePicker.cpp

```diff
@@ -320,3 +337,91 @@ connection CalendarDatePicker::connect_submit_signal(
     const SubmitSignal::slot_type& slot) const {
   return m_submit_signal.connect(slot);
 }
+
+bool CalendarDatePicker::eventFilter(QObject* watched, QEvent* event) {
+  if(event->type() == QEvent::KeyPress) {
+    auto e = static_cast<QKeyEvent*>(event);
+    if(watched == m_month_spinner &&
+        (e->key() == Qt::Key_Up || e->key() == Qt::Key_Down)) {
+      qApp->sendEvent(m_calendar_view, e);
+    } else {
+      auto current_index = m_calendar_view->get_current_model()->get_current();
+      if(current_index) {
+        if(*current_index == 0 && e->key() == Qt::Key_Left) {
+          m_model->set_current(*m_model->get_current() - days(1));
+          return true;
+        } else if(*current_index ==
+            m_calendar_view->get_list_model()->get_size() - 1 &&
+            e->key() == Qt::Key_Right) {
+          m_model->set_current(*m_model->get_current() + days(1));
+          return true;
+        }
+      }
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+boost::optional<int> CalendarDatePicker::get_index(date day) const {
+  for(auto i = 0; i < m_calendar_view->get_list_model()->get_size(); ++i) {
+    if(m_calendar_view->get_list_model()->get<date>(i) == day) {
+      return i;
+    }
+  }
+  return {};
+}
+
+void CalendarDatePicker::set_current_index(const optional<int>& index) {
+  auto current_block =
+    shared_connection_block(m_list_current_connection);
+  m_calendar_view->get_current_model()->set_current(index);
+}
+
+void CalendarDatePicker::on_current(const optional<date>& current) {
+  if(current) {
+    set_current_index(get_index(*current));
+  } else {
+    set_current_index({});
+  }
+}
+
+void CalendarDatePicker::on_current_month(date month) {
+  auto current_index = optional<int>();
+  auto list_has_focus = m_calendar_view->hasFocus() ||
+    m_calendar_view->isAncestorOf(focusWidget());
+  for(auto i = 0; i < m_calendar_view->get_list_model()->get_size(); ++i) {
+    auto minimum = m_model->get_minimum();
+    auto maximum = m_model->get_maximum();
+    auto current = m_calendar_view->get_list_model()->get<date>(i);
+    if(minimum && maximum) {
```

Comment:
I believe you can eliminate the test for ```minimum``` and ```maximum```. Review the following:

https://www.boost.org/doc/libs/1_77_0/libs/optional/doc/html/boost_optional/tutorial/relational_operators.html

A ```boost::none``` compares less than any other value.